### PR TITLE
Frontend: teach final module emitting jobs to dump a placeholder file for module semantic info

### PR DESCRIPTION
**Pull request by nkcsgexi**
This additional supplement output should capture semantic info the compiler has
captured while building a Swift module. Similar to the source info file, the content of
the semantic info file should only be consumed by local tooling written in Swift.

### CODE REVIEW

File: lib/FrontendTool/FrontendTool.cpp [hunk 1]
```diff
@@ -660,6 +660,22 @@ static void emitSwiftdepsForAllPrimaryInputsIfNeeded(
...
 }
 
+static bool writeModuleSemanticInfoIfNeeded(CompilerInstance &Instance) {
+  const auto &Invocation = Instance.getInvocation();
+  const auto &frontendOpts = Invocation.getFrontendOptions();
+  if (!frontendOpts.InputsAndOutputs.hasModuleSemanticInfoOutputPath())
+    return false;
+  std::error_code EC;
+  assert(frontendOpts.InputsAndOutputs.isWholeModule() &&
+         "TBDPath only makes sense when the whole module can be seen");
```
Comment (artemcm) [inline lib/FrontendTool/FrontendTool.cpp:670]:
```suggestion
         "Module Semantic Info path only makes sense when the whole module can be seen");
```

**Review by artemcm — APPROVED** [review]
Thanks! 👍🏼

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Pull request merged**